src/codegen/util/csv_scanner.cpp

```diff
@@ -0,0 +1,375 @@
+
+  const char *buf = buffer_;
+  uint32_t curr_buffer_pos = buffer_pos_;
+
+  while (true) {
+    if (curr_buffer_pos == buffer_end_) {
+      // We need to read more data from the CSV file. But first, we need to copy
+      // all the data in the read-buffer (i.e., [buffer_begin_, buffer_end_] to
+      // the line-buffer.
+      if (buffer_pos_ < curr_buffer_pos) {
+        AppendToLineBuffer(buffer_ + buffer_pos_,
+                           curr_buffer_pos - buffer_pos_);
+        buffer_pos_ = curr_buffer_pos;
+      }
+
+      // Reset positions
+      curr_buffer_pos = 0;
+
+      // Now, read more data
+      if (!NextBuffer()) {
+        // We hit en EOF
+        break;
+      }
+    }
+
+    // Read character
+    char c = buf[curr_buffer_pos++];
+
+    if (in_quote && c == escape) {
+      last_was_escape = !last_was_escape;
+    }
+    if (c == quote && !last_was_escape) {
+      in_quote = !in_quote;
+    }
+    if (c != escape) {
+      last_was_escape = false;
+    }
+
+    // Process the new-line character. If we a new-line and we're not currently
+    // in a quoted section, we're done.
+    if (c == '\n' && !in_quote) {
+      break;
+    }
+  }
+
+  // Flush remaining valid bytes
+  if (buffer_pos_ < curr_buffer_pos) {
+    AppendToLineBuffer(buffer_ + buffer_pos_, curr_buffer_pos - buffer_pos_);
+    buffer_pos_ = curr_buffer_pos;
+  }
+
+  // Increment line number
+  line_number_++;
+
+  // If we didn't transfer any bytes to the line buffer, we must have reached an
+  // EOF. If so, return null indicating there are no more lines.
+  if (line_len_ == 0) {
+    return nullptr;
+  }
+
+  // A full line has been transferred to the line buffer, but we also copied the
+  // newline character. Strip it off now.
+  line_len_--;
+  line_[line_len_] = '\0';
+
+  // Done
+  return line_;
+}
+
+void CSVScanner::ProduceCSV(char *line) {
+  const char delimiter = delimiter_;
+  const char quote = quote_;
+  const char escape = escape_;
+
+  // The iterator over characters in the line
+  char *iter = line;
+
+  for (uint32_t col_idx = 0; col_idx < num_cols_; col_idx++) {
+    char *col_begin = iter;
+    char *col_end = nullptr;
+
+    // We need to move col_end to the end of the column's data. Along the way,
+    // we may need to shift data down due to quotes and escapes. Inspired by
+    // Postgres.
+    {
+      char *out = col_begin;
+      while (true) {
+        // This first loop looks for either the delimiter character or the end
+        // of the line, indicating the end of a columns data. It breaks out of
+        // the loop if a quote character is found. It flows into a second loop
+        // whose only purpose is to find the end of the quoted section.
+        while (true) {
+          char c = *iter++;
+
+          // If we see the delimiter character, or the end of the string,
+          // finish
+          if (c == delimiter || c == '\0') {
+            col_end = out;
+            iter--;
```

Comment:
Why is `iter` decremented here? The next column loop iteration will read the delimiter or null-byte again, will it?

Reply:
The invariant we want to hold is that when this loop exists, `iter` either points to a newline character or a delimiter. We decrement here because we incremented earlier on line 300. Happy to take suggestions on how to make this clearer. CSV parsing is tricky :sweat: